pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -206,29 +244,35 @@ func (c *EtcdCertSignerController) createSecretForNode(node *corev1.Node, record
 		return err
 	}
 
-	nodeInternalIPs, err := dnshelpers.GetInternalIPAddressesForNodeName(node)
+	// create the certificates and update them in the API
+	pCert, pKey, err := tlshelpers.CreatePeerCertKey(etcdCASecret.Data["tls.crt"], etcdCASecret.Data["tls.key"], nodeInternalIPs)
 	if err != nil {
 		return err
 	}
 
-	// create the certificates and update them in the API
-	pCert, pKey, err := tlshelpers.CreatePeerCertKey(etcdCASecret.Data["tls.crt"], etcdCASecret.Data["tls.key"], nodeInternalIPs)
-	err = c.createSecret(etcdPeerClientCertName, pCert, pKey, recorder)
-	if err != nil {
+	if err := c.createSecret(etcdPeerClientCertName, pCert, pKey, recorder); err != nil {
 		return err
 	}
 	sCert, sKey, err := tlshelpers.CreateServerCertKey(etcdCASecret.Data["tls.crt"], etcdCASecret.Data["tls.key"], nodeInternalIPs)
-	err = c.createSecret(etcdServingCertName, sCert, sKey, recorder)
 	if err != nil {
 		return err
 	}
+
+	if err := c.createSecret(etcdServingCertName, sCert, sKey, recorder); err != nil {
+		return err
+	}
+
 	metricCert, metricKey, err := tlshelpers.CreateMetricCertKey(etcdMetricCASecret.Data["tls.crt"], etcdMetricCASecret.Data["tls.key"], nodeInternalIPs)
-	err = c.createSecret(metricsServingCertName, metricCert, metricKey, recorder)
 	if err != nil {
 		return err
 	}
 
-	return nil
+	if err := c.createSecret(metricsServingCertName, metricCert, metricKey, recorder); err != nil {
+		return err
+	}
+
+	// Return error to flag the mismatch of SAN to raise a degraded condition.
+	return utilerrors.NewAggregate(errors)
```

Comment:
not sure how comfortable I am with rolling new certificates by default in the face of SAN mismatch. Too many corner cases we are not covering.

Scenario 1. If cluster upgrades and on reboot node receives a new IP address.
Scenario 2. Machine is deleted and new machine with same name is created with new IP.

The first we would want to inform the user of the issue so they could resolve it. Would this change solve that problem? If we tolerate the IP change which could have implications on other components. Regardless we fail because membership was not changed to reflect new peerURLs.

Given the first problem, I am not so sure we can reasonably conclude that we should rotate certs with the information that this controller has.

Reply:
Pushed another update. The new code does the follow:

1. If any of the secrets are not found, generate new secrets
2. If all the secrets are found, but do not match SAN, then, raise the error but DO NOT generate new certs.
3. If all the secrets are found, and match the SAN, then, do nothing.